### DISCUSSION TITLE
How do Unity game loop phases and Entitas Redux system phases relate?

### DISCUSSION THREAD

**Discussion by JesseTG** [jeffcampbellmakesgames/Entitas-Redux#66] [Q&A] [answered]
Entitas Redux execution phases are intended to coincide with [Unity's game loop](https://docs.unity3d.com/Manual/ExecutionOrder). For the most part, their places within the game loop are easy to figure out:

- [`IFixedUpdateSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_fixed_update_system) is meant for [`MonoBehaviour.FixedUpdate`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.FixedUpdate)
- [`IInitializeSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_initialize_system) is meant for `MonoBehaviour`'s [`Start`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.Start), [`Awake`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.Awake), or [`OnEnable`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnEnable).
- [`ILateUpdateSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_late_update_system) is meant for [`MonoBehaviour.LateUpdate`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.LateUpdate)
- [`ITearDownSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_tear_down_system) is meant for `MonoBehaviour`'s [`OnDestroy`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnDestroy) or [`OnDisable`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnDisable)
- [`IUpdateSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_update_system) is meant for [`MonoBehaviour.Update`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.Update)

The intended sequencing of [`IReactiveSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_reactive_system) and [`ICleanupSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_cleanup_system) are less clear, however. In vanilla Entitas, `IReactiveSystem`s run alongside `IExecuteSystem`s and `ICleanupSystem`s can run any time after that frame's `IExecuteSystem`s. However, the new execution phases offered by Entitas Redux makes this question more complicated.

**When are systems of these types intended to be executed, relative to the others?**

**Comment by TheHacky** [comment]
I would always go with this order.

IUpdateSystems
IReactiveSystems
ICleanupSystems

Because I want to execute always before I react on changes and at the end I cleanup everything. If you this in Update or FixedUpdate depends on the game. If it Physics related I would do it in the FixedUpdate or have two different Features. One for Update and one for FixedUpdate.

**Comment by jeffcampbellmakesgames** [comment, marked as answer]
> Entitas Redux execution phases are intended to coincide with [Unity's game loop](https://docs.unity3d.com/Manual/ExecutionOrder). For the most part, their places within the game loop are easy to figure out:
> 
> * [`IFixedUpdateSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_fixed_update_system) is meant for [`MonoBehaviour.FixedUpdate`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.FixedUpdate)
> * [`IInitializeSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_initialize_system) is meant for `MonoBehaviour`'s [`Start`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.Start), [`Awake`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.Awake), or [`OnEnable`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnEnable).
> * [`ILateUpdateSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_late_update_system) is meant for [`MonoBehaviour.LateUpdate`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.LateUpdate)
> * [`ITearDownSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_tear_down_system) is meant for `MonoBehaviour`'s [`OnDestroy`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnDestroy) or [`OnDisable`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.OnDisable)
> * [`IUpdateSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_update_system) is meant for [`MonoBehaviour.Update`](https://docs.unity3d.com/ScriptReference/MonoBehaviour.Update)
> 

This is all dead-on for the most part. `IInitializeSystem` and `ITearDownSystem` don't necessarily have a well-defined Unity lifecycle method to associate with because how a game's systems are constructed and deconstructed depend a great deal on the app's overall architecture. 

For example, if the controller for a game's systems are scene-based, `Awake/OnDestroy` might be the best to use in that scenario. If they are controlled on a singleton that would live beyond scenes it's more likely there would be bespoke logic to that game that would determine when systems should be constructed and deconstructed.

I personally don't advocate for using `OnEnable`/`OnDisable` since most of the game systems I would write would use either of the two ways described above and/or other data-oriented ways of determining when to be active/inactive, but there shouldn't be anything that prohibits that.

> The intended sequencing of [`IReactiveSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_reactive_system) and [`ICleanupSystem`](https://jeffcampbellmakesgames.github.io/Entitas-Redux-API/interface_j_c_m_g_1_1_entitas_redux_1_1_i_cleanup_system) are less clear, however. In vanilla Entitas, `IReactiveSystem`s run alongside `IExecuteSystem`s and `ICleanupSystem`s can run any time after that frame's `IExecuteSystem`s. However, the new execution phases offered by Entitas Redux makes this question more complicated.

> 
> **When are systems of these types intended to be executed, relative to the others?**

I'd recommend reactive systems execute in the `Update` method along with update systems; this should lead to the same type of behaviour previously seen in Entitas and is how the example content is setup to demonstrate how to setup a systems controller.

Cleanup systems I would typically place after Update, Reactive, and LateUpdate as there could be data created by all three that would require cleaning up.

See the example controller setup [here](https://github.com/jeffcampbellmakesgames/Entitas-Redux/blob/d1c74ac6f96c37fb9ca57ef9edfa31d5b9a265c6/Unity/Assets/ExampleContent/Scenes/Visual%20Debug%20Example/VisualDebuggingExampleSystemsController.cs) for an example Systems setup.

```csharp
private void Update()
{
	_systems.Update();
	_systems.Execute();
}

private void LateUpdate()
{
	_systems.LateUpdate();
	_systems.Cleanup();
}
```
